torch/autograd/functional.py

```diff
@@ -429,6 +467,85 @@ def jacobian(func, inputs, create_graph=False, strict=False):
+        # NOTE: [Computing jacobian with vmap and grad for multiple outputs]
+        #
+        # Let's consider f(x) = (x**2, x.sum()) and let x = torch.randn(3).
+        # It turns out we can compute the jacobian of this function with a single
+        # call to autograd.grad by using vmap over the correct grad_outputs.
+        #
+        # Firstly, one way to compute the jacobian is to stack x**2 and x.sum()
+        # into a 4D vector. E.g., use g(x) = torch.stack([x**2, x.sum()])
+        #
+        # To get the first row of the jacobian, we call
+        # >>> autograd.grad(g(x), x, grad_outputs=torch.tensor([1, 0, 0, 0]))
+        # To get the 2nd row of the jacobian, we call
+        # >>> autograd.grad(g(x), x, grad_outputs=torch.tensor([0, 1, 0, 0]))
+        # and so on.
+        #
+        # Using vmap, we can vectorize all 4 of these computations into one by
+        # passing the standard basis for R^4 as the grad_output.
+        # vmap(partial(autograd.grad, g(x), x))(torch.eye(4)).
+        #
+        # Now, how do we compute the jacobian *without stacking the output*?
+        # We can just split the standard basis across the outputs. So to
+        # compute the jacobian of f(x), we'd use
+        # >>> autograd.grad(f(x), x, grad_outputs=_construct_standard_basis_for(...))
+        # The grad_outputs looks like the following:
+        # ( torch.tensor([[1, 0, 0],
+        #                 [0, 1, 0],
+        #                 [0, 0, 1],
+        #                 [0, 0, 0]]),
+        #   torch.tensor([[0],
+        #                 [0],
+        #                 [0],
+        #                 [1]]) )
+        #
+        # But we're not done yet!
+        # >>> vmap(partial(autograd.grad(f(x), x, grad_outputs=...)))
+        # returns a Tensor of shape [4, 3]. We have to remember to split the
+        # jacobian of shape [4, 3] into two:
+        # - one of shape [3, 3] for the first output
+        # - one of shape [   3] for the second output
+
+        # Step 1: Construct grad_outputs by splitting the standard basis
+        output_numels = tuple(output.numel() for output in outputs)
+        grad_outputs = _construct_standard_basis_for(outputs, output_numels)
+        flat_outputs = tuple(output.reshape(-1) for output in outputs)
+
+        # Step 2: Call vmap + autograd.grad
+        def vjp(grad_output):
+            vj = list(_autograd_grad(flat_outputs, inputs, grad_output, create_graph=create_graph))
+            for el_idx, vj_el in enumerate(vj):
+                if vj_el is not None:
+                    continue
+                vj[el_idx] = torch.zeros_like(inputs[el_idx])
+            return tuple(vj)
+
+        jacobians_of_flat_output = vmap(vjp)(grad_outputs)
+
+        # Step 3: The returned jacobian is one big tensor per input. In this step,
+        # we split each Tensor by output.
+        jacobian = tuple(
+            tuple(jac_out.view(output.shape + inp.shape)
```

Comment:
shouldn't the view be `view(output.shape, inp.shape)`?

Reply:
There are two ways to use view:
- pass var-args ints: tensor.view(1, 2, 3, 4, 5)
- pass a list/tuple of ints: tensor.view([1, 2, 3, 4, 5])

`jac_out.view(output.shape + input.shape)` is the latter, but if we wanted to do the former we could do `view(*output.shape, *inp.shape)`

Reply:
Ho these are tuples, not numbers. My bad!